#devtools::install_github("DevonDeRaad/SNPfiltR")
library(SNPfiltR)
library(vcfR)
library(adegenet)
library(hierfstat)
library(pegas)
library(poppr)
library(PopGenReport)
library(ggplot2)

##############################################################
# DENOVO OUTPUT
##############################################################

setwd("D:/RADseq/nodiflorum")
denovo_assembly = vcfR::read.vcfR("populations_output/populations.snps.vcf")
denovo_assembly

pops_all = vcfR::read.vcfR("populations_output/populations.all.vcf")

#denovo_assembly_genlight = vcfR::vcfR2genlight(denovo_assembly)
#denovo_assembly_genlight

#denovo_assembly_genind = vcfR::vcfR2genind(denovo_assembly)

popmap = data.frame(id=colnames(denovo_assembly@gt)[2:length(colnames(denovo_assembly@gt))],pop=substr(colnames(denovo_assembly@gt)[2:length(colnames(denovo_assembly@gt))], 3,11))

#visualize distributions
SNPfiltR::hard_filter(vcfR=denovo_assembly)

denovo_assembly = SNPfiltR::hard_filter(vcfR=denovo_assembly, depth = 5, gq = 35)
denovo_assembly = SNPfiltR::filter_allele_balance(denovo_assembly)
# dashed line indicates a mean depth across all SNPs of 77.4
SNPfiltR::max_depth(denovo_assembly)

denovo_assembly = SNPfiltR::max_depth(denovo_assembly, maxdepth = 125)
denovo_assembly

SNPfiltR::missing_by_sample(vcfR=denovo_assembly, popmap = popmap)

# 13 samples are above a 0.9 missing data cutoff, and were removed from VCF
denovo_assembly = SNPfiltR::missing_by_sample(vcfR=denovo_assembly, cutoff = .9)

#subset popmap to only include retained individuals
popmap = popmap[popmap$id %in% colnames(denovo_assembly@gt),]

#remove invariant sites generated by dropping individuals
denovo_assembly = SNPfiltR::min_mac(denovo_assembly, min.mac = 1)

miss = SNPfiltR::assess_missing_data_pca(vcfR=denovo_assembly, 
        popmap = popmap, thresholds = .8, clustering = FALSE)

# Picking joint bandwidth of 0.0632
SNPfiltR::missing_by_snp(denovo_assembly)

miss = SNPfiltR::assess_missing_data_pca(vcfR=denovo_assembly, popmap = popmap, 
                  thresholds = c(.6, .65, .7), clustering = FALSE)

denovo_assembly = SNPfiltR::missing_by_snp(denovo_assembly, cutoff = .80)

denovo_assembly

denovo_assembly.mac = SNPfiltR::min_mac(vcfR = denovo_assembly, min.mac = 2)

#assess clustering without MAC cutoff
miss = SNPfiltR::assess_missing_data_tsne(denovo_assembly, popmap, 
                                          clustering = FALSE)
#assess clustering with MAC cutoff
miss = SNPfiltR::assess_missing_data_tsne(denovo_assembly.mac, popmap, 
                                          clustering = FALSE)

#plot depth per snp and per sample
dp = vcfR::extract.gt(denovo_assembly, element = "DP", as.numeric=TRUE)
vcfR::heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq = vcfR::extract.gt(denovo_assembly, element = "GQ", as.numeric=TRUE)
vcfR::heatmap.bp(gq, rlabels = FALSE)

denovo_assembly

# original file: 83 samples, 86,274 SNPs
# after filtering: 70 samples, 4,862 SNPs
# ~18X smaller

vcfR::write.vcf(denovo_assembly, "populations_output/populations.snps.filtered.vcf")

#linkage filter vcf to thin SNPs to one per 500bp
denovo_assembly.thin = SNPfiltR::distance_thin(denovo_assembly, min.distance = 500)
vcfR::write.vcf(denovo_assembly.thin, "populations_output/populations.snps.filtered.thinned.vcf")

# > denovo_assembly
# ***** Object of Class vcfR *****
#   70 samples
# 1334 CHROMs
# 4,862 variants
# Object size: 23.6 Mb
# 8.612 percent missing data
# *****        *****         *****
#   
#   > denovo_assembly.thin
# ***** Object of Class vcfR *****
#   70 samples
# 1334 CHROMs
# 1,334 variants
# Object size: 6.6 Mb
# 10.68 percent missing data
# *****        *****         *****

denovo_assembly = vcfR::read.vcfR("populations_output/populations.snps.filtered.vcf")

# convert to genind
denovo_assembly_genind = vcfR::vcfR2genind(denovo_assembly)
denovo_assembly_genind

# convert to genlight
denovo_assembly_genlight = vcfR::vcfR2genlight(denovo_assembly)
denovo_assembly_genlight@ind.names

# Extract the grouping code (1â€“2 uppercase letters from the beginning)
# populations by country
country_pops = sub("^([A-Z]+).*", "\\1", denovo_assembly_genlight@ind.names)
country_pops

broad_pops = ifelse(country_pops == "SA", "native",
                    ifelse(country_pops %in% c("US", "MX"), "invaded",
                    "introduced"))

broad_pops

# check
cbind(country_pops, broad_pops)

country_pops
broad_pops
denovo_assembly_genlight@ind.names

country_pop_definitions = as.data.frame(cbind(denovo_assembly_genlight@ind.names, 
                                              country_pops) )
colnames(country_pop_definitions) = c("id", "pop_id")
write.table(country_pop_definitions, "post_filtering_data/fastStructure/country_pop_defs.txt",
            row.names = F, quote = F)

broad_pop_definitions = as.data.frame( cbind(denovo_assembly_genlight@ind.names, 
                                             broad_pops) )
colnames(broad_pop_definitions) = c("id", "pop_id")
write.table(broad_pop_definitions, "post_filtering_data/fastStructure/broad_pop_defs.txt",
            row.names = F, quote = F)

# create copies of the original data, but assign different population structures

# broad grouping
broad_grouping.genlight = denovo_assembly_genlight
broad_grouping.genlight@pop = as.factor(broad_pops)

################################################################################
# save a version for SplitsTree
sample.div = StAMPP::stamppNeisD(broad_grouping.genlight, pop = FALSE)
#export for splitstree
StAMPP::stamppPhylip(distance.mat=sample.div, 
                     file="Rscripts/post_filtering_data/nodiflorum_splits.txt")
################################################################################

broad_grouping.genind = denovo_assembly_genind
broad_grouping.genind@pop = as.factor(broad_pops)

# country grouping
country_grouping.genlight = denovo_assembly_genlight
country_grouping.genlight@pop = as.factor(country_pops)

country_grouping.genind = denovo_assembly_genind
country_grouping.genind@pop = as.factor(country_pops)

broad_grouping.genlight
country_grouping.genlight

broad_grouping.genind
country_grouping.genind

#####################################################################################
# create STRUCTURE files

dartR::gl2faststructure(x = broad_grouping.genlight, outpath=getwd(),
              outfile = "post_filtering_data/fastStructure/broad/broad_structure.str")

dartR::gl2faststructure(x = country_grouping.genlight, outpath=getwd(),
              outfile = "post_filtering_data/fastStructure/country/country_structure.str")

######################################################################################

myInset <- function(dapc){
  temp <- dapc$pca.eig
  temp <- 100* cumsum(temp)/sum(temp)
  plot(temp, col=rep(c("black","lightgrey"),
                     c(dapc$n.pca,1000)), ylim=c(0,100),
       xlab="PCA axis", ylab="Cumulated variance (%)",
       cex=1, pch=20, type="h", lwd=2)
}

broad_myCol <- c("gold", "red", "forestgreen")

broad.dapc = adegenet::dapc(broad_grouping.genlight, var.contrib = TRUE, 
               scale = FALSE, n.pca = NULL, n.da = nPop(broad_grouping.genlight) - 1)

ade4::scatter(broad.dapc, clabel = 0, legend = TRUE, cell=0,
              cstar = 0, cex = 4, pch = 20, solid = 0.4, scree.da=FALSE,
              col = broad_myCol, bg="white",posi.pca="bottomleft",
              cleg = 0.75, xax = 1, yax = 2)

add.scatter(myInset(broad.dapc), posi="bottomright",
            inset=c(-0.03,-0.08), ratio=.28,
            bg=transp("white"))

#ade4::scatter(broad.dapc, 1,1, col = broad_myCol, scree.da = FALSE, legend = TRUE, solid = 0.4)

as.factor(country_pops)
country_myCol <- c("gold", "deeppink", "black", "purple", "royalblue", "darkorange",
                   "red", "forestgreen", "lightblue", "darkred")

country.dapc = adegenet::dapc(country_grouping.genlight, var.contrib = TRUE, 
                 scale = FALSE, n.pca = 40, n.da = nPop(country_grouping.genlight) - 1)
ade4::scatter(country.dapc, posi.pca = "topleft", clabel = 0, leg = TRUE,
              cstar = 0, cell = 0, solid = 0.6, pch = 20, cex = 3, scree.da=FALSE,
              col = country_myCol)

add.scatter(myInset(country.dapc), posi="bottomright",
            inset=c(-0.03,-0.12), ratio=.28,
            bg=transp("white"))


##################################
# population statistics
##################################

##########################
# get broad pop stats
##########################

broad_poppr_result = poppr::poppr(broad_grouping.genind)
broad_poppr_result <- broad_poppr_result[-nrow(broad_poppr_result), ] # remove last "total" row

country_poppr_result = poppr::poppr(country_grouping.genind)
country_poppr_result <- country_poppr_result[-nrow(country_poppr_result), ]

#############
# add more stats to the above DFs

broad.stats = hierfstat::basic.stats(data = broad_grouping.genind)
country.stats = hierfstat::basic.stats(data = country_grouping.genind)

# add to existing DFs
broad_poppr_result$Fis = colMeans(broad.stats$Fis, na.rm = TRUE)
broad_poppr_result$Ho = colMeans(broad.stats$Ho, na.rm = TRUE)
broad_poppr_result$Hs =colMeans(broad.stats$Hs, na.rm = TRUE) 

country_poppr_result$Fis = colMeans(country.stats$Fis, na.rm = TRUE)
country_poppr_result$Ho = colMeans(country.stats$Ho, na.rm = TRUE)
country_poppr_result$Hs = colMeans(country.stats$Hs, na.rm = TRUE)

# find private alleles and add to existing DFs

broad_private_alleles = poppr::private_alleles(broad_grouping.genind, report = "data.frame",
                                               level = "population", count.alleles = FALSE) 

broad_total_private_perpop = aggregate(count ~ population, 
                                       data = broad_private_alleles, sum) 

country_private_alleles = poppr::private_alleles(country_grouping.genind, report = "data.frame",
                                                 level = "population", count.alleles = FALSE) 

country_total_private_perpop = aggregate(count ~ population, 
                                         data = country_private_alleles, sum)

broad_poppr_result$private_alleles = broad_total_private_perpop$count
country_poppr_result$private_alleles = country_total_private_perpop$count

write.csv(broad_poppr_result, 
          "Rscripts/post_filtering_data/broad_poppr_results.csv", row.names = FALSE)
write.csv(country_poppr_result, 
          "Rscripts/post_filtering_data/country_poppr_results.csv", row.names = FALSE)


# FST values
broad.FST = hierfstat::genet.dist(broad_grouping.genind, method = "WC84")
broad.FST = as.data.frame(as.matrix(broad.FST))
write.csv(broad.FST, "Rscripts/post_filtering_data/broadFST.csv", row.names = T)

country.FST = hierfstat::genet.dist(country_grouping.genind, method = "WC84")
country.FST = as.data.frame(as.matrix(country.FST))
write.csv(country.FST, "Rscripts/post_filtering_data/countryFST.csv", row.names = T)

#################################
# plotting
#################################

broad_poppr_result = read.csv("post_filtering_data/broad_poppr_results.csv") %>%
  dplyr::select(-c(File, Hexp, SE, E.5, eMLG)) %>%
  dplyr::rename("private alleles" = "private_alleles") %>%
  tidyr::pivot_longer(cols = -Pop, names_to = "statistic", values_to = "value") 
head(broad_poppr_result)

country_poppr_result = read.csv("post_filtering_data/country_poppr_results.csv") %>%
  dplyr::select(-c(File, Hexp, SE, E.5, eMLG)) %>%
  dplyr::rename("private alleles" = "private_alleles") %>%
  tidyr::pivot_longer(cols = -Pop, names_to = "statistic", values_to = "value")
head(country_poppr_result)


popstats.broad.plot = ggplot(broad_poppr_result, aes(x = Pop, y = value, fill = Pop)) +
  geom_bar(stat = "identity", show.legend = FALSE, alpha = 0.6, , colour = "black") +
  facet_wrap(~statistic, scales = "free_y") +
  scale_fill_manual(values = c("goldenrod", "darkred", "forestgreen")) +  # Custom fill colors
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(x = "Population", y = "Value", title = "") ;popstats.broad.plot

ggsave("post_filtering_data/popstats.broad.png", popstats.broad.plot, width = 6.2, dpi=400,
       height = 5, units = "in")


country_poppr_result$fill_color <- ifelse(grepl("\\bSA\\b", country_poppr_result$Pop), "forestgreen",
                                          ifelse(grepl("\\b(US|MX)\\b", country_poppr_result$Pop), "darkred", "goldenrod"))

popstats.country.plot = ggplot(country_poppr_result, aes(x = Pop, y = value, 
                          fill = fill_color)) +
  geom_bar(stat = "identity", show.legend = FALSE, alpha = 0.6, colour ="black") +
  facet_wrap(~statistic, scales = "free_y") +
  scale_fill_identity() +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(x = "Population", y = "Value", title = "") ;popstats.country.plot

ggsave("post_filtering_data/popstats.country.png", popstats.country.plot, width = 8, dpi=400,
       height = 5, units = "in")


broad_grouping.genpop = adegenet::genind2genpop(broad_grouping.genind)

